File: src/incident_triage_agent/workflow.py
from __future__ import annotations

from loguru import logger

from .domain import Scenario, TriageRun, WorkflowState
from .llm import LLMClient
from .policy import evaluate_safety
from .scoring import score_run
from .tools import MockOperationalTools


log = logger.bind(component="workflow")


class TriageWorkflow:
    def __init__(self, tools: MockOperationalTools, llm_client: LLMClient) -> None:
        self.tools = tools
        self.llm_client = llm_client

    def run(self, scenario: Scenario) -> TriageRun:
        log.info("Starting workflow for scenario '{}'.", scenario.name)
        run = TriageRun(scenario=scenario)
        run.transition(WorkflowState.RECEIVED)
        log.debug("State transition: {}.", WorkflowState.RECEIVED.value)

        run.evidence_package = self.tools.build_evidence_package(scenario)
        run.transition(WorkflowState.CONTEXT_GATHERED)
        log.debug("State transition: {}.", WorkflowState.CONTEXT_GATHERED.value)

        run.transition(WorkflowState.LLM_DECISION_REQUESTED)
        log.debug("State transition: {}.", WorkflowState.LLM_DECISION_REQUESTED.value)
        run.validation = self.llm_client.decide(run.evidence_package)

        if not run.validation.valid or not run.validation.decision:
            run.transition(WorkflowState.RECOVERABLE_FAILURE)
            log.warning("State transition: {}.", WorkflowState.RECOVERABLE_FAILURE.value)
            run.transition(WorkflowState.SCORED)
            log.debug("State transition: {}.", WorkflowState.SCORED.value)
            run.scorecard = score_run(run)
            return run

        run.transition(WorkflowState.DECISION_VALIDATED)
        log.debug("State transition: {}.", WorkflowState.DECISION_VALIDATED.value)
        run.safety = evaluate_safety(run.validation.decision, run.evidence_package)

        if run.safety.status == "approval_required":
            run.transition(WorkflowState.APPROVAL_PENDING)
            log.debug("State transition: {}.", WorkflowState.APPROVAL_PENDING.value)
            run.transition(WorkflowState.SIMULATED_ACTION_RECORDED)
            log.debug("State transition: {}.", WorkflowState.SIMULATED_ACTION_RECORDED.value)
        elif run.safety.status == "needs_human_input":
            run.transition(WorkflowState.HUMAN_INPUT_NEEDED)
            log.debug("State transition: {}.", WorkflowState.HUMAN_INPUT_NEEDED.value)
        else:
            run.transition(WorkflowState.VERIFICATION_READY)
            log.debug("State transition: {}.", WorkflowState.VERIFICATION_READY.value)

        run.transition(WorkflowState.SCORED)
        log.debug("State transition: {}.", WorkflowState.SCORED.value)
        run.scorecard = score_run(run)
        log.info("Workflow complete for scenario '{}'.", scenario.name)
        return run

File: src/workflow.ts
import type { Scenario, WorkflowState } from "./domain";
import type { EvidencePackage, InvestigationStep, MockOperationalTools } from "./evidence";
import type { ExplanationValidation, LLMDecisionClient, TriageExplanation, ValidationResult } from "./llm";
import { noopLogger, type TriageLogger } from "./logger";
import { evaluateSafety, type SafetyResult } from "./policy";
import { scoreRun, type Scorecard } from "./scoring";

export type RunStatus = "running" | "completed" | "recoverable_failure";

export interface InvestigationTrace {
  summary: string;
  steps: InvestigationStep[];
}

export interface TriageRun {
  runId: string;
  runStatus: RunStatus;
  scenario: Scenario;
  states: WorkflowState[];
  investigation?: InvestigationTrace;
  evidencePackage?: EvidencePackage;
  validation?: ValidationResult;
  explanation?: TriageExplanation;
  explanationValidation?: ExplanationValidation;
  safety?: SafetyResult;
  scorecard?: Scorecard;
}

export class TriageWorkflow {
  constructor(
    private readonly tools: MockOperationalTools,
    private readonly llmClient: LLMDecisionClient,
    private readonly logger: TriageLogger = noopLogger,
  ) {}

  async run(scenario: Scenario): Promise<TriageRun> {
    this.logger.info({ component: "workflow", scenario: scenario.name }, "Starting workflow");
    const run: TriageRun = { runId: `triage-run:${scenario.name}`, runStatus: "running", scenario, states: [] };

    this.transition(run, "received");
    run.evidencePackage = this.tools.buildEvidencePackage(scenario);
    run.investigation = buildInvestigationTrace(run.evidencePackage);
    this.logger.info({
      component: "tools",
      scenario: scenario.name,
      evidenceCount: run.evidencePackage.evidence.length,
      missingContextCount: run.evidencePackage.missingContext.length,
    }, "Evidence package ready");
    this.transition(run, "context_gathered");

    this.transition(run, "llm_decision_requested");
    run.validation = await this.llmClient.decide(run.evidencePackage);
    if (run.validation.explanation) {
      run.explanation = run.validation.explanation;
    }
    if (run.validation.explanationValidation) {
      run.explanationValidation = run.validation.explanationValidation;
    }

    if (!run.validation.valid || !run.validation.decision) {
      this.logger.warn({ component: "llm", errors: run.validation.errors }, "LLM decision validation failed");
      this.transition(run, "recoverable_failure");
      this.transition(run, "scored");
      run.scorecard = scoreRun(run);
      run.runStatus = "recoverable_failure";
      this.logger.info({ component: "scoring", passed: passedCount(run.scorecard), total: totalCount(run.scorecard) }, "Scorecard complete");
      return run;
    }

    this.logger.info({
      component: "llm",
      incidentClass: run.validation.decision.incidentClass,
      nextAction: run.validation.decision.nextAction,
      confidence: run.validation.decision.confidence,
    }, "LLM decision validated");
    this.transition(run, "decision_validated");
    run.safety = evaluateSafety(run.validation.decision, run.evidencePackage);
    this.logger.info({
      component: "policy",
      status: run.safety.status,
      approvalRequired: run.safety.approvalRequired,
    }, "Safety evaluated");

    if (run.safety.status === "approval_required") {
      this.transition(run, "approval_pending");
      this.transition(run, "simulated_action_recorded");
    } else if (run.safety.status === "needs_human_input") {
      this.transition(run, "human_input_needed");
    } else {
      this.transition(run, "verification_ready");
    }

    this.transition(run, "scored");
    run.scorecard = scoreRun(run);
    run.runStatus = "completed";
    this.logger.info({ component: "scoring", passed: passedCount(run.scorecard), total: totalCount(run.scorecard) }, "Scorecard complete");
    this.logger.info({ component: "workflow", scenario: scenario.name }, "Workflow complete");
    return run;
  }

  private transition(run: TriageRun, state: WorkflowState): void {
    run.states.push(state);
    this.logger.debug({ component: "workflow", state }, "State transition");
  }
}

function passedCount(scorecard: Scorecard): number {
  return Object.values(scorecard.scores).filter(Boolean).length;
}

function totalCount(scorecard: Scorecard): number {
  return Object.values(scorecard.scores).length;
}

function buildInvestigationTrace(evidencePackage: EvidencePackage): InvestigationTrace {
  const found = evidencePackage.investigationSteps.filter((step) => step.status === "found").length;
  const notFound = evidencePackage.investigationSteps.filter((step) => step.status === "not_found").length;
  const skipped = evidencePackage.investigationSteps.filter((step) => step.status === "skipped").length;
  const errored = evidencePackage.investigationSteps.filter((step) => step.status === "error").length;
  const statusParts = [
    `${found} found`,
    notFound > 0 ? `${notFound} not found` : undefined,
    skipped > 0 ? `${skipped} skipped` : undefined,
    errored > 0 ? `${errored} error` : undefined,
  ].filter((part): part is string => part !== undefined);
  return {
    summary: `Collected ${evidencePackage.evidence.length} evidence item(s) across ${evidencePackage.investigationSteps.length} investigation step(s): ${statusParts.join(", ")}.`,
    steps: evidencePackage.investigationSteps,
  };
}
